sirfilesystem.c

```diff
@@ -157,6 +157,75 @@ char* _sir_getcwd(void) {
 #endif
 }
 
+#if defined(__OpenBSD__)
+static int
+_sir_openbsdself(char* out, int capacity, int* dirname_length)
+{
+  char   buffer1[4096];
+  char   buffer2[PATH_MAX];
+  char   buffer3[PATH_MAX];
+  char** argv     = (char**)buffer1;
+  char*  resolved = NULL;
+  int    length   = -1;
+
+  for (;;) {
+    int mib[4] = { CTL_KERN, KERN_PROC_ARGS, getpid(), KERN_PROC_ARGV };
+    size_t size;
+
+    if (sysctl(mib, 4, NULL, &size, NULL, 0) != 0) break;
+    if (size > sizeof(buffer1)) {
+      argv = (char**)malloc(size);
+      if (!argv) break;
+    }
+
+    if (sysctl(mib, 4, argv, &size, NULL, 0) != 0) break;
+    if (strchr(argv[0], '/')) {
+      resolved = realpath(argv[0], buffer2);
+      if (!resolved) break;
+    } else {
+      const char* PATH = getenv("PATH");
+      if (!PATH) break;
+      size_t argv0_length = strlen(argv[0]);
+      const char* begin = PATH;
+      while (1) {
+        const char* separator = strchr(begin, ':');
+        const char* end = separator ? separator : begin + strlen(begin);
+        if (end - begin > 0) {
+          if (*(end - 1) == '/') --end;
+          if (((end - begin) + 1UL + argv0_length + 1UL) <= sizeof(buffer2)) {
+            memcpy(buffer2, begin, end - begin);
+            buffer2[end - begin] = '/';
+            memcpy(buffer2 + (end - begin) + 1, argv[0], argv0_length + 1);
+            resolved = realpath(buffer2, buffer3);
+            if (resolved) break;
+          }
+        }
+        if (!separator) break;
+        begin = ++separator;
+      }
+      if (!resolved) break;
+    }
+
+    length = (int)strlen(resolved);
+    if (length <= capacity) {
+      memcpy(out, resolved, (unsigned long)length);
+      if (dirname_length) {
+        int i;
+        for (i = length - 1; i >= 0; --i) {
```

Comment:
What is this doing? Setting a bunch of the output to '/'? How come?

Reply:
When it iterates through the PATH it sets / if it's missing, so looking for "prog" and given  "/usr/bin:/bin" we check for "/usr/bin/prog" and "/bin/prog" (instead of /binprog").